client/asset/eth/eth.go

```diff
@@ -507,68 +719,282 @@ func (eth *ExchangeWallet) createFundingCoin(amount uint64) *coin {
 // In UTXO based coins, the returned []dex.Bytes contains the redeem scripts for the
 // selected coins, but since there are no redeem scripts in Ethereum, nil is returned.
 // Equal number of coins and redeem scripts must be returned.
-func (eth *ExchangeWallet) FundOrder(ord *asset.Order) (asset.Coins, []dex.Bytes, error) {
-	if eth.gasFeeLimit < ord.DEXConfig.MaxFeeRate {
+func (w *AssetWallet) FundOrder(ord *asset.Order) (asset.Coins, []dex.Bytes, error) {
+	cfg := ord.DEXConfig
+
+	if w.gasFeeLimit < cfg.MaxFeeRate {
 		return nil, nil, fmt.Errorf(
 			"%v: server's max fee rate %v higher than configured fee rate limit %v",
-			ord.DEXConfig.Symbol,
-			ord.DEXConfig.MaxFeeRate,
-			eth.gasFeeLimit)
-	}
-	maxSwapFees := ord.DEXConfig.MaxFeeRate * ord.DEXConfig.SwapSize * ord.MaxSwapCount
-	refundFees := dexeth.RefundGas(0) * eth.gasFeeLimit
-	fundsNeeded := ord.Value + maxSwapFees + refundFees
-	coins := asset.Coins{eth.createFundingCoin(fundsNeeded)}
-	err := eth.lockFunds(fundsNeeded)
+			cfg.Symbol,
+			cfg.MaxFeeRate,
+			w.gasFeeLimit)
+	}
+
+	g, err := w.gasEstimate(int(ord.MaxSwapCount), cfg, ord.RedeemConfig)
 	if err != nil {
+		return nil, nil, fmt.Errorf("error estimating swap gas: %v", err)
+	}
+
+	ethToLock := cfg.MaxFeeRate * g.Swap * ord.MaxSwapCount
+	// Note: In a future refactor, we could lock the redemption funds here too
+	// and signal to the user so that they don't call `RedeemN`. This has the
+	// same net effect, but avoids a lockFunds -> unlockFunds for us and likely
+	// some work for the caller as well. We can't just always do it that way and
+	// remove RedeemN, since we can't guarantee that the redemption asset is in
+	// our fee-family. though it could still be an AccountRedeemer.
+	//
+	// ethToLock += biggestUint64(g.oneRefund, g.oneRedeem) * cfg.MaxFeeRate * ord.MaxSwapCount
+
+	ethWallet := w
+	var success bool
+	var coin asset.Coin
+	if w.assetID == BipID {
+		ethToLock += ord.Value
+		coin = w.createFundingCoin(ethToLock)
+	} else {
+		// Check if we need to up the allowance.
+		ethWallet = w.token.parent
+		currentAllowance, err := w.tokenAllowance()
+		if err != nil {
+			return nil, nil, fmt.Errorf("error retrieving current allowance: %w", err)
+		}
+
+		if currentAllowance.Cmp(unlimitedAllowanceReplenishThreshold) < 0 {
+			ethBal, err := ethWallet.balance()
+			if err != nil {
+				return nil, nil, fmt.Errorf("error getting eth balance: %w", err)
+			}
+			approveGas, err := w.approveGas(unlimitedAllowance, cfg)
+			if err != nil {
+				return nil, nil, fmt.Errorf("error estimating allowance gas: %w", err)
+			}
+			if (approveGas*cfg.MaxFeeRate)+ethToLock > ethBal.Available {
+				return nil, nil, fmt.Errorf("parent balance %d doesn't cover contract approval (%d) and tx fees (%d)",
+					ethBal.Available, approveGas*cfg.MaxFeeRate, ethToLock)
+			}
+			if _, err := w.approveToken(unlimitedAllowance, cfg.MaxFeeRate, cfg.Version); err != nil {
```

Comment:
I haven't read the whole PR yet, but I'm assuming that on the first sell trade of the token, this is where the approval would happen. If this is the case then there would need to be additional handling on the server, because this trade cannot yet be booked. Having the unlimited approval as a manual step completely avoids this issue. I know you think this is a bad user experience, but anyone who has used an erc20 token is completely familiar with this step. It's just an extra click before you sell the token.

Also, I think these checks of the allowance going below 2^256/2 are not necessary. Other than maybe some absolute ridiculous token, there will be no tokens anywhere near 2^256 max supply, and even for that ridiculous token, the current user would have had to initiate swaps with the entire supply of the token before their allowance would run out, so I don't think we need to worry about that.

Reply:
> there would need to be additional handling on the server, because this trade cannot yet be booked

Can you expand on this? Doesn't the nonce order guarantee that the approval will be processed first.

From what I can tell, we're not checking the client's allowance server-side, and I didn't think we had any plans to do that. I was just going to relegate it to the client as part of configuration. Nothing we could do can prevent a client from sending an `approve(contractAddr, 0)` through after we accept their order anyway. 
The cost of approval is next to nothing, compartively, and correctly reported in the pre-order estimates.

If we did want to check allowance server-side, we would have to add latency handling though, which is what I think you were getting at.

To add manual approval, we would need to add a new UI dialog, new endpoint in webserver, new exported method from `*Core`, new exported method from `eth.ExchangeWallet`, and a new method or specialty interface in **interface.go**. And then we need to maintain all of that forever. And that's so we can force a user to click a button that they have no choice but to click in order to use the wallet that they just created. We also have to explain why we are forcing them to separately click this button to use the wallet. Why would they click "Create" if they were then going to say no to configuring the wallet for use? It makes for a bad UX, imo. Approval is not optional. It's part of trading with this asset. We don't need to pass the burden on to the user.

Reply:
Hm I'm probably  biased because all the other apps I've used have had approval as a manual step.  I think you're right, we don't need to do any validation on the server side. 

I think the ideal way would be the approval happens automatically when the user is doing their first buy order.  In that case this code that does the approval should be in `ReserveN` instead of `FundOrder`.